subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/BaseInstrumentingArtifactTransform.java

```diff
@@ -140,13 +114,25 @@ private static String getOutputPath(File input, Function<String, String> instrum
         return INSTRUMENTED_DIR_NAME + "/" + instrumentedEntryNameMapper.apply(input.getName());
     }
 
-    protected abstract InterceptorTypeRegistryAndFilter provideInterceptorTypeRegistryAndFilter();
-
-
-    private boolean isAgentSupported() {
-        return getParameters().getAgentSupported().get();
+    protected void doOutputOriginalArtifact(File input, TransformOutputs outputs) {
+        // Output original file if it's safe to load from cache loader ELSE copy an entry
+        if (input.isDirectory()) {
+            // Directories are ok to use outside the cache, since they are not locked by the daemon.
+            outputs.dir(input);
+        } else if (internalServices.get().getGlobalCacheLocations().isInsideGlobalCache(input.getAbsolutePath())) {
+            // Jars that are already in the global cache don't need to be copied, since
+            // the global caches are additive only and jars shouldn't be deleted or changed during the build.
+            outputs.file(input);
+        } else {
+            // Jars that are in some mutable location (e.g. build/ directory) need to be copied to the global cache,
+            // since daemon keeps them locked when loading them to a classloader, which prevents e.g. deleting the build directory on windows
+            File copyOfOriginalFile = outputs.file(ORIGINAL_DIR_NAME + "/" + input.getName());
+            GFileUtils.copyFile(input, copyOfOriginalFile);
+        }
     }
```

Comment:
🤔 This looks very brittle, depending on implementation details to make this decision. I wonder how we could make this more future proof...

Reply:
I would need some more context, what do you mean with `implementation details` here?